p2pserver/protocols/utils/cons_peers.go

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package utils
+
+import (
+	"bytes"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/ontio/ontology-crypto/keypair"
+	"github.com/ontio/ontology/common"
+	"github.com/ontio/ontology/common/config"
+	vconfig "github.com/ontio/ontology/consensus/vbft/config"
+	"github.com/ontio/ontology/core/ledger"
+	"github.com/ontio/ontology/smartcontract/service/native/governance"
+	"github.com/ontio/ontology/smartcontract/service/native/utils"
+)
+
+type GovNodeResolver interface {
+	IsGovNode(key keypair.PublicKey) bool
+}
+
+type GovNodeMockResolver struct {
+	govNode map[string]struct{}
+}
+
+func NewGovNodeMockResolver(gov []string) *GovNodeMockResolver {
+	govNode := make(map[string]struct{}, len(gov))
+	for _, node := range gov {
+		govNode[node] = struct{}{}
+	}
+
+	return &GovNodeMockResolver{govNode}
+}
+
+func (self *GovNodeMockResolver) IsGovNode(key keypair.PublicKey) bool {
+	pubKey := vconfig.PubkeyID(key)
+	_, ok := self.govNode[pubKey]
+
+	return ok
+}
+
+type GovNodeLedgerResolver struct {
+	db   *ledger.Ledger
+	view uint32
+
+	cache unsafe.Pointer // atomic pointer to GovCache, avoid read&write data race
```

Comment:
why this?

Reply:
gov node will not change in the same view, so add a cache to avoid query the ledger every time.